File: golang/cdkPipelines/inspectionStack.go
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

// Permission is hereby granted, free of charge, to any person obtaining a copy of this
// software and associated documentation files (the "Software"), to deal in the Software
// without restriction, including without limitation the rights to use, copy, modify,
// merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
// permit persons to whom the Software is furnished to do so.

// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
// INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
// HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
// SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

package cdkPipelines

import (
	"fmt"
	"strings"

	"github.com/aws/aws-cdk-go/awscdk/v2"
	ec2 "github.com/aws/aws-cdk-go/awscdk/v2/awsec2"
	iam "github.com/aws/aws-cdk-go/awscdk/v2/awsiam"
	lambda "github.com/aws/aws-cdk-go/awscdk/v2/awslambda"
	logs "github.com/aws/aws-cdk-go/awscdk/v2/awslogs"
	nf "github.com/aws/aws-cdk-go/awscdk/v2/awsnetworkfirewall"
	cr "github.com/aws/aws-cdk-go/awscdk/v2/customresources"
	"github.com/aws/constructs-go/constructs/v10"
	"github.com/aws/jsii-runtime-go"
)

type NetworkFirewallStackProps struct {
	awscdk.StackProps
}

func NetworkFirewallStack(scope constructs.Construct, id string, cidr string, orgCidr string, transitGWId *string, props *NetworkFirewallStackProps) {

	var sprops awscdk.StackProps
	if props != nil {
		sprops = props.StackProps
	}
	stack := awscdk.NewStack(scope, &id, &sprops)

	vpc := ec2.NewVpc(stack, jsii.String("InspectionVPC"), &ec2.VpcProps{
		IpAddresses: ec2.IpAddresses_Cidr(&cidr),
		SubnetConfiguration: &[]*ec2.SubnetConfiguration{
			{
				Name:       jsii.String("Tgw_Subnet"),
				SubnetType: ec2.SubnetType_PRIVATE_ISOLATED,
				CidrMask:   jsii.Number(26),
			},
			{
				Name:       jsii.String("Firewall_Subnet"),
				SubnetType: ec2.SubnetType_PRIVATE_WITH_EGRESS,
				CidrMask:   jsii.Number(27),
			},
			{
				Name:       jsii.String("Public"),
				SubnetType: ec2.SubnetType_PUBLIC,
				CidrMask:   jsii.Number(27),
			},
		},
	})

	tGWSubnetIDs := vpc.SelectSubnets(&ec2.SubnetSelection{
		SubnetGroupName: jsii.String("Tgw_Subnet"),
	}).SubnetIds

	tGWAttachment := ec2.NewCfnTransitGatewayAttachment(stack, jsii.String("TGW_Attachment"), &ec2.CfnTransitGatewayAttachmentProps{
		TransitGatewayId: transitGWId,
		SubnetIds:        tGWSubnetIDs,
		VpcId:            vpc.VpcId(),
		Options:          ec2.CfnTransitGatewayAttachment_OptionsProperty{ApplianceModeSupport: jsii.String("enable")},
		Tags: &[]*awscdk.CfnTag{
			{
				Key:   jsii.String("routeTable"),
				Value: jsii.String("inspection"),
			},
		},
	})

	ec2.NewCfnTransitGatewayRoute(stack, jsii.String("TGW_Route"), &ec2.CfnTransitGatewayRouteProps{
		DestinationCidrBlock:       jsii.String("0.0.0.0/0"),
		TransitGatewayAttachmentId: tGWAttachment.AttrId(),
		TransitGatewayRouteTableId: awscdk.Fn_ImportValue(jsii.String("WorkloadRouteTableId")),
	})

	firewallRules := NetworkFirewallRules(stack, "NetworkFirewallRules", nil)

	fwSubnets := vpc.SelectSubnetObjects(&ec2.SubnetSelection{
		SubnetGroupName: jsii.String("Firewall_Subnet"),
	})

	var fwSubnetList []*nf.CfnFirewall_SubnetMappingProperty
	for _, privateSub := range *fwSubnets {
		subnetMProps := &nf.CfnFirewall_SubnetMappingProperty{
			SubnetId: privateSub.SubnetId(),
		}
		fwSubnetList = append(fwSubnetList, subnetMProps)
	}

	networkFw := nf.NewCfnFirewall(stack, jsii.String("Network_Firewall"), &nf.CfnFirewallProps{
		FirewallName:      jsii.String("EgressInspectionFirewall"),
		FirewallPolicyArn: firewallRules.FwPolicyArn(),
		SubnetMappings:    fwSubnetList,
		VpcId:             vpc.VpcId(),
	})

	fwFlowLogsGroup := logs.NewLogGroup(stack, jsii.String("FWFlowLogsGroup"), &logs.LogGroupProps{
		LogGroupName:  jsii.String("NetworkFirewallFlowLogs"),
		RemovalPolicy: awscdk.RemovalPolicy_DESTROY,
	})

	fwAlertLogsGroup := logs.NewLogGroup(stack, jsii.String("FWAlertLogsGroup"), &logs.LogGroupProps{
		LogGroupName:  jsii.String("NetworkFirewallAlertLogs"),
		RemovalPolicy: awscdk.RemovalPolicy_DESTROY,
	})

	nf.NewCfnLoggingConfiguration(stack, jsii.String("FirewallLoggingConfig"), &nf.CfnLoggingConfigurationProps{
		FirewallArn: networkFw.Ref(),
		LoggingConfiguration: nf.CfnLoggingConfiguration_LoggingConfigurationProperty{
			LogDestinationConfigs: []interface{}{
				&nf.CfnLoggingConfiguration_LogDestinationConfigProperty{
					LogDestination: map[string]*string{
						"logGroup": fwFlowLogsGroup.LogGroupName(),
					},
					LogDestinationType: jsii.String("CloudWatchLogs"),
					LogType:            jsii.String("FLOW"),
				},
				&nf.CfnLoggingConfiguration_LogDestinationConfigProperty{
					LogDestination: map[string]*string{
						"logGroup": fwAlertLogsGroup.LogGroupName(),
					},
					LogDestinationType: jsii.String("CloudWatchLogs"),
					LogType:            jsii.String("ALERT"),
				},
			},
		},
	})

	RouteLambdaRole := iam.NewRole(stack, jsii.String("routeLambdaRole"), &iam.RoleProps{
		AssumedBy: iam.NewServicePrincipal(jsii.String("lambda.amazonaws.com"), nil),
		Path:      jsii.String("/"),
		ManagedPolicies: &[]iam.IManagedPolicy{
			iam.ManagedPolicy_FromAwsManagedPolicyName(jsii.String("service-role/AWSLambdaBasicExecutionRole")),
		},
	})

	RouteLambdaRole.AddToPolicy(
		iam.NewPolicyStatement(&iam.PolicyStatementProps{
			Actions: jsii.Strings("network-firewall:DescribeFirewall"),
			Effect:  iam.Effect_ALLOW,
			Resources: &[]*string{
				networkFw.AttrFirewallArn(),
			},
		}),
	)

	tGWSubnets := vpc.SelectSubnetObjects(&ec2.SubnetSelection{
		SubnetGroupName: jsii.String("Tgw_Subnet"),
	})

	var tgwSubsIds []*string

	for _, subnet := range *tGWSubnets {
		subId := subnet.SubnetId()
		tgwSubsIds = append(tgwSubsIds, subId)
	}

	var cloudWanSubnetArns []*string

	for _, tgwSub := range *tGWSubnets {
		rtArns := awscdk.Arn_Format(&awscdk.ArnComponents{
			Service:  jsii.String("ec2"),
			Resource: jsii.String(fmt.Sprintf("route-table/%s", *tgwSub.RouteTable().RouteTableId())),
		},
			stack,
		)
		cloudWanSubnetArns = append(cloudWanSubnetArns, rtArns)
	}

	pubSubnets := vpc.SelectSubnetObjects(&ec2.SubnetSelection{
		SubnetGroupName: jsii.String("Public"),
	})

	var pubSubsIds []*string

	for _, subnet := range *pubSubnets {
		subId := subnet.SubnetId()
		pubSubsIds = append(pubSubsIds, subId)
	}

	var pubSubnetArns []*string
	for _, pubSub := range *pubSubnets {
		rtArns := awscdk.Arn_Format(&awscdk.ArnComponents{
			Service:  jsii.String("ec2"),
			Resource: jsii.String(fmt.Sprintf("route-table/%s", *pubSub.RouteTable().RouteTableId())),
		},
			stack,
		)
		pubSubnetArns = append(pubSubnetArns, rtArns)
	}

	listArns := append(cloudWanSubnetArns, pubSubnetArns...)

	RouteLambdaRole.AddToPolicy(
		iam.NewPolicyStatement(&iam.PolicyStatementProps{
			Actions:   jsii.Strings("ec2:CreateRoute", "ec2:DeleteRoute"),
			Effect:    iam.Effect_ALLOW,
			Resources: &listArns,
		}),
	)

	customRouteLambda := lambda.NewFunction(stack, jsii.String("RoutesFunction"), &lambda.FunctionProps{
		Runtime: lambda.Runtime_PYTHON_3_9(),
		Handler: jsii.String("index.lambda_handler"),
		Role:    RouteLambdaRole,
		Timeout: awscdk.Duration_Seconds(jsii.Number(60)),
		Code:    lambda.Code_FromAsset(jsii.String("lambda/routes"), nil),
	})

	customResource := cr.NewProvider(stack, jsii.String("provider"), &cr.ProviderProps{
		OnEventHandler: customRouteLambda,
		LogRetention:   logs.RetentionDays_ONE_DAY,
	})

	// TODO: not copied over the dependency on the Lambda as it should be
	// implicit; verify.

	// Create a default route towards AWS Network firewall endpoints. Select
	// all subnets in group CloudWANAttacment and use custom lambda function
	// to find AWS Network Firewall endpoint that is in same availability
	// zone as subnet.

	subs := vpc.SelectSubnetObjects(&ec2.SubnetSelection{
		SubnetGroupName: jsii.String("Tgw_Subnet"),
	})

	// Create a custom resource for each TGw subnet
	for _, subnet := range *subs {
		// Create CloudFormation custom resource to update firewall routing
		awscdk.NewCustomResource(stack, jsii.String(fmt.Sprintf("FirewallRoute-%s", strings.SplitN(*subnet.Node().Path(), "/", 1))), &awscdk.CustomResourceProps{
			ServiceToken: customResource.ServiceToken(),
			Properties: &map[string]interface{}{
				"FirewallArn":     networkFw.AttrFirewallArn(),
				"SubnetAz":        subnet.AvailabilityZone(),
				"RouteTableId":    subnet.RouteTable().RouteTableId(),
				"DestinationCidr": "0.0.0.0/0",
			},
		})
	}

	pubSubs := vpc.SelectSubnetObjects(&ec2.SubnetSelection{
		SubnetGroupName: jsii.String("Public"),
	})

	// Create a custom resource for each Public subnet
	for _, subnet := range *pubSubs {
		// Create CloudFormation custom resource to update firewall routing
		awscdk.NewCustomResource(stack, jsii.String(fmt.Sprintf("ReturnRoute-%s", strings.SplitN(*subnet.Node().Path(), "/", 1))), &awscdk.CustomResourceProps{
			ServiceToken: customResource.ServiceToken(),
			Properties: &map[string]interface{}{
				"FirewallArn":     networkFw.AttrFirewallArn(),
				"SubnetAz":        subnet.AvailabilityZone(),
				"RouteTableId":    subnet.RouteTable().RouteTableId(),
				"DestinationCidr": orgCidr,
			},
		})
	}

	fwSubs := vpc.SelectSubnetObjects(&ec2.SubnetSelection{
		SubnetGroupName: jsii.String("Firewall_Subnet"),
	})

	// Create a Route for each FW subnet
	for _, subnet := range *fwSubs {
		// Create CloudFormation custom resource to update firewall routing
		ec2.NewCfnRoute(stack, jsii.String(fmt.Sprintf("OrganisationRoute-%s", strings.SplitN(*subnet.Node().Path(), "/", 1))), &ec2.CfnRouteProps{
			RouteTableId:         subnet.RouteTable().RouteTableId(),
			DestinationCidrBlock: &orgCidr,
			TransitGatewayId:     transitGWId,
		}).AddDependency(tGWAttachment)
	}
}

File: typescript/cdkPipelines/lib/inspectionStack.ts
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

// Permission is hereby granted, free of charge, to any person obtaining a copy of this
// software and associated documentation files (the "Software"), to deal in the Software
// without restriction, including without limitation the rights to use, copy, modify,
// merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
// permit persons to whom the Software is furnished to do so.

// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
// INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
// HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
// SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

import * as cdk from 'aws-cdk-lib';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import { Construct } from 'constructs';
import { CustomResource, Duration, RemovalPolicy } from 'aws-cdk-lib';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as nodejs from 'aws-cdk-lib/aws-lambda-nodejs';
import * as logs from 'aws-cdk-lib/aws-logs';
import * as nf from 'aws-cdk-lib/aws-networkfirewall';
import * as cr from 'aws-cdk-lib/custom-resources';
import { NetworkFirewallRules } from '../lib/firewallRules';
import { ORGANISATION_CIDR } from '../lib/configurations';

interface NetworkFirewallProps extends cdk.StackProps {
  cidr: string;
  orgCidr: string;
  transitGWId: string;
  workloadRouteTableID: string;
}

export class NetworkFirewallStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: NetworkFirewallProps) {
    super(scope, id, props);

    const vpc = new ec2.Vpc(this, 'InspectionVPC', {
      ipAddresses: ec2.IpAddresses.cidr(props.cidr),
      subnetConfiguration: [
        {
          name: 'TGWSubnets',
          subnetType: ec2.SubnetType.PRIVATE_ISOLATED,
          cidrMask: 26,
        },
        {
          name: 'FirewallSubnets',
          subnetType: ec2.SubnetType.PRIVATE_WITH_EGRESS,
          cidrMask: 27,
        },
        {
          name: 'Public',
          subnetType: ec2.SubnetType.PUBLIC,
          cidrMask: 27,
        },
      ],
    });

    const tGWSubnets = vpc.selectSubnets({
      subnetGroupName: 'TGWSubnets',
    });

    const pubSubnets = vpc.selectSubnets({
      subnetGroupName: 'Public',
    });

    const tGWAttachment = new ec2.CfnTransitGatewayAttachment(
      this,
      'TGWAttachment',
      {
        transitGatewayId: props.transitGWId,
        subnetIds: tGWSubnets.subnetIds,
        vpcId: vpc.vpcId,
        options: {
          applianceModeSupport: 'ENABLED',
        },
        tags: [
          {
            key: 'routeTable',
            value: 'inspection',
          },
        ],
      }
    );

    new ec2.CfnTransitGatewayRoute(this, 'TGWRoute', {
      destinationCidrBlock: '0.0.0.0/0',
      transitGatewayAttachmentId: tGWAttachment.attrId,
      transitGatewayRouteTableId: props.workloadRouteTableID,
    });

    const firewallRules = new NetworkFirewallRules(
      this,
      'NetworkFirewallRules'
    );

    const fwSubnets = vpc.selectSubnets({
      subnetGroupName: 'FirewallSubnets',
    });

    let subnetList: nf.CfnFirewall.SubnetMappingProperty[] = [];
    for (const subnet of fwSubnets.subnets) {
      const subnetMappingProperty: nf.CfnFirewall.SubnetMappingProperty = {
        subnetId: subnet.subnetId,
      };
      subnetList.push(subnetMappingProperty);
    }

    const networkFw = new nf.CfnFirewall(this, 'NetworkFirewall', {
      firewallName: 'EgressInspectionFirewall',
      firewallPolicyArn: firewallRules.fwPolicy.attrFirewallPolicyArn,
      subnetMappings: subnetList,
      vpcId: vpc.vpcId,
    });

    const fwFlowLogsGroup = new logs.LogGroup(this, 'FwFlowLogsGroup', {
      logGroupName: 'NetworkFirewallFlowLogs',
      removalPolicy: RemovalPolicy.DESTROY,
    });

    const fwAlertLogGroup = new logs.LogGroup(this, 'FWAlertLogsGroup', {
      logGroupName: 'NetworkFirewallAlertLogs',
      removalPolicy: RemovalPolicy.DESTROY,
    });

    new nf.CfnLoggingConfiguration(this, 'FirewallLoggingConfg', {
      firewallArn: networkFw.ref,
      loggingConfiguration: {
        logDestinationConfigs: [
          {
            logDestination: {
              logGroup: fwFlowLogsGroup.logGroupName,
            },
            logDestinationType: 'CloudWatchLogs',
            logType: 'FLOW',
          },
          {
            logDestination: {
              logGroup: fwAlertLogGroup.logGroupName,
            },
            logDestinationType: 'CloudWatchLogs',
            logType: 'ALERT',
          },
        ],
      },
    });

    const routeLambdaRole = new iam.Role(this, 'RouteLambdaRole', {
      assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
      managedPolicies: [
        iam.ManagedPolicy.fromAwsManagedPolicyName(
          'service-role/AWSLambdaBasicExecutionRole'
        ),
      ],
    });

    routeLambdaRole.addToPolicy(
      new iam.PolicyStatement({
        effect: iam.Effect.ALLOW,
        actions: ['network-firewall:DescribeFirewall'],
        resources: [networkFw.attrFirewallArn],
      })
    );

    const customRouteLambda = new nodejs.NodejsFunction(this, 'endpoint', {
      functionName: 'CustomRouteLambda',
      role: routeLambdaRole,
      timeout: Duration.seconds(30),
    });

    const provider = new cr.Provider(this, 'Provider', {
      onEventHandler: customRouteLambda,
      logRetention: logs.RetentionDays.ONE_DAY,
    });

    vpc
      .selectSubnets({ subnetGroupName: 'Public' })
      .subnets.forEach((subnet) => {
        const subnetName = subnet.node.path.split('/').pop(); // E.g. TransitGatewayStack/InspectionVPC/PublicSubnet1

        // Custom resource returns AWS Network Firewall endpoint ID in correct availability zone.
        const endpoint = new CustomResource(
          this,
          `AnfEndpointFor-${subnetName}`,
          {
            serviceToken: provider.serviceToken,
            properties: {
              FirewallName: networkFw.firewallName,
              AvailabilityZone: subnet.availabilityZone,
            },
          }
        );

        // Create default route towards firewall endpoint from Public subnets.
        new ec2.CfnRoute(this, `${subnetName}AnfRoute`, {
          destinationCidrBlock: ORGANISATION_CIDR,
          routeTableId: subnet.routeTable.routeTableId,
          vpcEndpointId: endpoint.getAttString('EndpointId'),
        });
      });

    vpc
      .selectSubnets({ subnetGroupName: 'TGWSubnets' })
      .subnets.forEach((subnet) => {
        const subnetName = subnet.node.path.split('/').pop(); // E.g. TransitGatewayStack/InspectionVPC/PublicSubnet1

        // Custom resource returns AWS Network Firewall endpoint ID in correct availability zone.
        const endpoint = new CustomResource(
          this,
          `AnfEndpointFor-${subnetName}`,
          {
            serviceToken: provider.serviceToken,
            properties: {
              FirewallName: networkFw.firewallName,
              AvailabilityZone: subnet.availabilityZone,
            },
          }
        );

        // Create default route towards firewall endpoint from TGW subnets.
        new ec2.CfnRoute(this, `${subnetName}AnfRoute`, {
          destinationCidrBlock: '0.0.0.0/0',
          routeTableId: subnet.routeTable.routeTableId,
          vpcEndpointId: endpoint.getAttString('EndpointId'),
        });
      });

    vpc
      .selectSubnets({ subnetGroupName: 'FirewallSubnets' })
      .subnets.forEach((subnet) => {
        const subnetName = subnet.node.path.split('/').pop(); // E.g. TransitGatewayStack/InspectionVPC/PublicSubnet1

        // Create route towards organisation network from firewall subnets.
        new ec2.CfnRoute(this, `${subnetName}AnfRoute`, {
          destinationCidrBlock: ORGANISATION_CIDR,
          routeTableId: subnet.routeTable.routeTableId,
          transitGatewayId: props.transitGWId,
        });
      });
  }
}
